Added features that allow consumers to customise View Location and runtime stylesheet loading

BenWhite27 opened the pull request:
Added `IBlockPreviewRequestEnricher` service that allows enrichment of the http context for runtime view and stylesheet location.

Modified `BlockPreviewService` to allow extensibility around `CreateViewData` and `GetViewResult`.  Making use of a new `BlockPreviewContext` class to pass around the data needed to make decisions in custom logic.  Also added `BlockPreviewService.GetStylesheetPath` to allow customisation of the stylesheet that is delivered at runtime.

Updated the Backoffice UI code to obtain the stylesheet via API call.

Also added some example implementations to the CMS project.

This is all in relation to #115 with the goal of allowing a package consumer to display theme specific views in the back office.  In theory the theme could be specified as a property in the content node hierarchy or based on the domain configured on the root content node.

rickbutterfield commented:
Hey @BenWhite27! I haven't forgotten about this, hoping to get on to it very soon. In the meantime, please can you pull in the latest changes from `v3/dev` just to make sure it's up to date?

rickbutterfield commented:
Hey @BenWhite27, I've finally got round to looking at this and I'm a big fan! It's a great, simple implementation and I can see how useful it would be for different use cases, plus how it could be extended in the future to open up other functionality that's currently locked away in `BlockPreviewService`.

I have run your branch locally as well as trying to run it against `v5/dev` and all appears to work well in my very basic attempt at matching things. I know you have pulled it up-to-date once already but if you can do it against `v5/dev` I'd be happy to merge it in and get it released to coincide with the Umbraco 17 version.

BenWhite27 commented:
Great news @rickbutterfield I'll give it a shot this weekend if I can find the time.

BenWhite27 committed:
Remove unused throwOnError parameter for ConvertToElement
Updated nullability for the method and all references.

BenWhite27 committed:
Created a BlockPreviewContext object
to pass between methods that offer opportunities for 3rd party customisation.
Set those methods as protected virtual to allow customisation.

BenWhite27 committed:
Add a GetStylesheetPath method to the BlockPreviewService

BenWhite27 committed:
Add request enrichment to BlockPreviewApiController

Updated BlockPreviewApiController to include a new
IBlockPreviewRequestEnricher field and modified the
constructor to initialize it.

Introduced IBlockPreviewRequestEnricher
interface and a NoopBlockPreviewRequestEnricher.

BenWhite27 committed:
Removed use of RuntimeCache for the GetPublishedContent.

BenWhite27 committed:
Add actions for style sheet paths to BlockPreviewApiController

BenWhite27 committed:
Update Backoffice UI code to obtain Stylesheet path from the API.

BenWhite27 committed:
Provide example implementation and configuration

rickbutterfield committed:
Merge branch 'v5/dev' into v3/dev

BenWhite27 commented:
Hi @rickbutterfield 
Apologies I wasn't quite finished with the work on this one and should have converted the PR back to draft.  I'll try and get the rest done today, can it be merged again or will it need a new PR?

BenWhite27 commented:
HI @rickbutterfield 
I can confirm there's an issue with the code in this PR, I've already fixed it locally but there's a [github outage](https://www.githubstatus.com/incidents/5q7nmlxz30sk) preventing me from pushing it at the moment.  Will try again tonight and raise a new PR with the fixes based on the latest bits in v5/dev.  After that if you think it's useful I'd like to create another PR with a decent, polished example of how it can be used.
